### PR TITLE
E2E tests for Oracle

### PR DESCRIPTION
**Pull request by poojantcs**

### CODE REVIEW

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+#
+
+@Oracle
+Feature: Oracle sink - Verify Oracle sink plugin design time scenarios
+  @Oracle_Required
```
Comment (poojantcs) [inline oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature]: 1. We should mention source/sink via tags. Eg. @Oracle @Source
2. What is the significance of: @Oracle_Required ? Are we following some process here?

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  Scenario: To verify Oracle sink plugin validation errors for mandatory fields
```
Comment (poojantcs) [inline oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature]: Validation scenarios should be part of the Validation feature file

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    Then Replace input plugin property: "password" with value: "password" for Credentials and Authorization related fields
+    Then Select radio button plugin property: "role" with value: "sysdba"
+    Then Select radio button plugin property: "connectionType" with value: "service"
+    Then Enter input plugin property: "referenceName" with value: "targetRef"
+    Then Replace input plugin property: "database" with value: "databaseName"
+    Then Replace input plugin property: "tableName" with value: "targetTable"
+    Then Replace input plugin property: "defaultBatchValue" with value: "batchValue"
+    Then Validate "Oracle" plugin properties
+    Then Close the Plugin Properties page
+
```
Comment (poojantcs) [inline oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature]: Please remove the empty lines

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTimeValidation.feature [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    Then Verify Oracle plugin in-line error message for incorrect Reference Name: "oracleInvalidReference"
+
+  @Oracle_Required
+  Scenario: To verify Oracle sink plugin validation error message with invalid database
+    Given Open Datafusion Project to configure pipeline
+    When Expand Plugin group in the LHS plugins list: "Source"
+    When Select plugin: "Oracle" from the plugins list as: "Source"
+    When Expand Plugin group in the LHS plugins list: "Sink"
+    When Select plugin: "Oracle" from the plugins list as: "Sink"
+    Then Connect plugins: "Oracle" and "Oracle2" to establish connection
```
Comment (poojantcs) [inline oracle-plugin/src/e2e-test/features/sink/OracleDesignTimeValidation.feature:57]: We can update the reference name of the plugins to "Oracle_source" and "Oracle_sink" for better readability of the test

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature (hunk 1 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature]: Just added as it was mentioned in existing. Not required hence removed.

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature (hunk 2 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature]: These were design time phase field scenarios, and same was categorised in RTM. Please let me know if these scenarios are to merged into "OracleDesignTimeValidation.feature" file.

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature (hunk 3 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature]: Done.

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTimeValidation.feature (hunk 4 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/features/sink/OracleDesignTimeValidation.feature:57]: It is a framework step which actually takes the plugin title, so we cannot change it.

**Comment by itsmekumari** [comment]
Screenshot 2023-04-13 at 1 16 07 PM
Screenshot 2023-04-13 at 1 21 07 PM

Screenshot 2023-04-13 at 1 22 24 PM


Screenshot 2023-04-13 at 1 36 41 PM

Screenshot 2023-04-13 at 1 46 24 PM
Screenshot 2023-04-13 at 1 51 59 PM
Screenshot 2023-04-13 at 2 27 05 PM

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature (hunk 2 above)
Comment (Vipinofficial11) [reply oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature]: Yes this scenario needs to be in ``` OracleDesignTimeValidation ```. Whenever we are validating the plugin using different permutation and combination of properties those scenarios should be in DesignTime and whenever we are validating an error message that should be done in DesignTimeValidation.

File: oracle-plugin/src/e2e-test/features/source/OracleDesignTime.feature [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+#
+
+@Oracle
+Feature: Oracle source- Verify Oracle source plugin design time scenarios
+
+  Scenario: To verify Oracle source plugin validation errors for mandatory fields
```
Comment (Vipinofficial11) [inline oracle-plugin/src/e2e-test/features/source/OracleDesignTime.feature]: This scenario should be in Design time Validation file.

File: oracle-plugin/src/e2e-test/features/source/OracleRunTimeMacro.feature [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+    Then Enter runtime argument value "invalidPassword" for key "oraclePassword"
+    Then Enter runtime argument value "invalidDatabaseName" for key "oracleDatabase"
+    Then Run the preview of pipeline with runtime arguments
+    Then Wait till pipeline preview is in running state
+    Then Open and capture pipeline preview logs
+    Then Verify the preview run status of pipeline in the logs is "failed"
+    Then Close the pipeline logs
+    Then Close the preview
+
+  @ORACLE_SOURCE_TEST @ORACLE_SINK_TEST @Oracle_Required
```
Comment (Vipinofficial11) [inline oracle-plugin/src/e2e-test/features/source/OracleRunTimeMacro.feature]: Why are we using @Oracle_Required hook in our test cases. This concept is related to Github Actions when we only run some specific tests whenever a PR is raised to decrease the Build run time. Is it done in any of the tests written by google team ?

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks_BQ.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+import java.nio.file.Paths;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.NoSuchElementException;
+import java.util.UUID;
+
+/**
+ * GCP test hooks.
+ */
+public class TestSetupHooks_BQ {
```
Comment (Vipinofficial11) [inline oracle-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks_BQ.java]: We don't have to create a new class to write BQ hooks you can add these hooks in the TestSetupHook class.

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks_BQ.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+  @After(order = 1, value = "@BQ_SOURCE_TEST")
+  public static void deleteTempSourceBQTable() throws IOException, InterruptedException {
+    String bqSourceTable = PluginPropertyUtils.pluginProp("bqSourceTable");
+    BigQueryClient.dropBqQuery(bqSourceTable);
+    BeforeActions.scenario.write("BQ source Table " + bqSourceTable + " deleted successfully");
+    PluginPropertyUtils.removePluginProp("bqSourceTable");
+  }
+
+  private static void createSourceBQTableWithQueries(String bqCreateTableQueryFile, String bqInsertDataQueryFile)
```
Comment (Vipinofficial11) [inline oracle-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks_BQ.java]: This method is unused, also why do we need to create test data in the BQ table using 2 different ways. We can finalise one way and can use it overall. What do you think @poojantcs ?

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/oracle/actions/OracleActions.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import io.cdap.e2e.utils.PluginPropertyUtils;
+import io.cdap.e2e.utils.SeleniumDriver;
+import io.cdap.e2e.utils.SeleniumHelper;
+import io.cdap.e2e.utils.WaitHelper;
+import io.cdap.plugin.oracle.locators.OracleLocators;
+import io.cucumber.core.logging.Logger;
+import io.cucumber.core.logging.LoggerFactory;
+import org.junit.Assert;
+import org.openqa.selenium.Keys;
+import org.openqa.selenium.interactions.Actions;
```
Comment (Vipinofficial11) [inline oracle-plugin/src/e2e-test/java/io.cdap.plugin/oracle/actions/OracleActions.java]: Remove unused imports

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/oracle/stepsdesign/Oracle.java [hunk 10]
```diff
@@ -52,4 +57,35 @@ public void validateTheValuesOfRecordsTransferredToTargetTableIsEqualToTheValues
...
+  }
+
+  @Then("Verify Oracle plugin in-line error message for incorrect Table Name: {string}")
+  public void verifyOraclePluginInLineErrorMessageForIncorrectTableName(String fields) {
+    OracleActions.verifyGroupByPluginPropertyInlineErrorMessageForRow
+      ("tableName",
+       PluginPropertyUtils.errorProp(E2ETestConstants.ERROR_MSG_INVALID_TABLE_NAME)
+         .replace("TABLE", PluginPropertyUtils.pluginProp(fields)));
+    OracleActions.verifyGroupByPluginPropertyInlineErrorMessageForColor("tableName");
+  }
```
Comment (Vipinofficial11) [inline oracle-plugin/src/e2e-test/java/io.cdap.plugin/oracle/stepsdesign/Oracle.java:100]: Why are we writing these steps in here ? All the above steps are present in cdap-e2e-test repo. Reuse the steps from the framework and remove all the unnecessary Action,Steps and Locators.

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/utils/E2ETestConstants.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ * the License.
+ */
+package io.cdap.plugin.utils;
+
+/**
+ * E2E test constants.
+ */
+public class E2ETestConstants {
+  public static final String ERROR_MSG_INVALID_REFERENCE_NAME = "errorMessageOracleInvalidReferenceName";
+  public static final String ERROR_MSG_INVALID_TABLE_NAME = "errorMessageInvalidTableName";
```
Comment (Vipinofficial11) [inline oracle-plugin/src/e2e-test/java/io.cdap.plugin/utils/E2ETestConstants.java]: we will not be needing this class if we use common framework steps please update.

File: oracle-plugin/src/e2e-test/resources/pluginParameters.properties [hunk 12]
```diff
@@ -55,3 +55,28 @@ longVarcharColumns=(ID VARCHAR2(100) PRIMARY KEY, COL1 LONG VARCHAR)
...
+batchValue=40
+rowPrefetch=60
+invalidRef=invalidRef&^*&&*
+oracleInvalidReference=invalidRef&^*&&*
+oracleInvalidTable=invalidTable
+invalidDatabaseName=invalidDB%$^%*
+numberOfSplits=2
+zeroValue=0
+invalidImportQuery=select
+invalidTableNameImportQuery=select * from abc;
```
Comment (Vipinofficial11) [inline oracle-plugin/src/e2e-test/resources/pluginParameters.properties:82]: Separate the properties using comments.

File: oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature (hunk 2 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/features/sink/OracleDesignTime.feature]: Removed from OracleDesignTime.feature and added in OracleDesignTimeValidation.feature file. Will update the RTM

File: oracle-plugin/src/e2e-test/features/source/OracleDesignTime.feature (hunk 5 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/features/source/OracleDesignTime.feature]: Done. Will update the RTM.

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks_BQ.java (hunk 7 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks_BQ.java]: Removed and added in existing TestSetUpHooks class.

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks_BQ.java (hunk 8 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks_BQ.java]: As discussed, removed for now, but we may use this while creation of BQ tables with required query, while creating run time oracle sink scenarios.

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/oracle/actions/OracleActions.java (hunk 9 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/java/io.cdap.plugin/oracle/actions/OracleActions.java]: Removed class, as not required.

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/oracle/stepsdesign/Oracle.java (hunk 10 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/java/io.cdap.plugin/oracle/stepsdesign/Oracle.java:100]: Reused steps from framework.

File: oracle-plugin/src/e2e-test/java/io.cdap.plugin/utils/E2ETestConstants.java (hunk 11 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/java/io.cdap.plugin/utils/E2ETestConstants.java]: Done.

File: oracle-plugin/src/e2e-test/resources/pluginParameters.properties (hunk 12 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/resources/pluginParameters.properties:82]: Done.

File: oracle-plugin/src/e2e-test/features/source/OracleRunTimeMacro.feature (hunk 6 above)
Comment (itsmekumari) [reply oracle-plugin/src/e2e-test/features/source/OracleRunTimeMacro.feature]: Removed.

**Comment by Vipinofficial11** [comment]
Closing the PR as the external review is done and its already Approved.

**Pull request closed without merging**